public/app/features/alerting/unified/hooks/useProduceNewRuleGroup.tsx

```diff
@@ -0,0 +1,119 @@
+import { Action } from '@reduxjs/toolkit';
+import { useCallback, useState } from 'react';
+
+import { dispatch, getState } from 'app/store/store';
+import { RuleGroupIdentifier } from 'app/types/unified-alerting';
+import { RulerGrafanaRuleDTO, RulerRuleDTO, RulerRuleGroupDTO } from 'app/types/unified-alerting-dto';
+
+import { AlertGroupUpdated, alertRuleApi } from '../api/alertRuleApi';
+import { deleteRuleAction, pauseRuleAction, ruleGroupReducer } from '../reducers/ruler/ruleGroups';
+import { fetchRulesSourceBuildInfoAction, getDataSourceRulerConfig } from '../state/actions';
+
+type ProduceResult = RulerRuleGroupDTO | AlertGroupUpdated;
+
+// @TODO the manual state tracking here is not great, but I don't have a better idea that works /shrug
+function useProduceNewRuleGroup() {
```

Comment:
I think we could probably do with a comment here to explain why this is important/how it should be used?

```suggestion
/**
 * Hook for reuse that handles freshly fetching a rule group's definition, applying an action to it,
 * and then performing the API mutations necessary to persist the change.
 *
 * All rule groups changes should ideally be implemented as a wrapper around this hook,
 * to ensure that we always protect as best we can against accidentally overwriting changes,
 * and to guard against user concurrency issues.
 *
 * @TODO the manual state tracking here is not great, but I don't have a better idea that works /shrug
 */
 function useProduceNewRuleGroup() {
```